bdb-ip.md

```diff
@@ -0,0 +1,437 @@
+
+After releasing BDB v1.0, breaking changes to the external APIs are not
+permitted anymore. Additions to the APIs are however always allowed.
+
+To resolve the outstanding issues, we'll hence limit
+the scope of this document to the already existing endpoints.
+
+
+### First task
+
+Please check the two listed milestones above for the correctness of categorized
+issues.
+
+
+### Methodology
+
+This document will go address each of the issues in the [breaking
+changes milestone](https://github.com/bigchaindb/bigchaindb/milestone/11) and
+discuss possible solutions to them. Once there is a consensus on the solutions,
+we can implement them into BDB core (mainly HTTP API), py and js driver and
+the BDB CLI.
+
+
+### Issues in milestone
+
+#### [Inconsistent naming of short-hand ids](https://github.com/bigchaindb/bigchaindb/issues/1134)
+
+Current naming of ids according to [HTTP API documentation
+(master)](https://docs.bigchaindb.com/projects/server/en/master/http-client-server-api.html)
+and the [Websocket Event Stream API
+(master)](https://docs.bigchaindb.com/projects/server/en/master/websocket-event-stream-api.html):
+
+- Transaction id:
+    - `tx_id` in query parameters and Event Stream API
+    - `txid` in `transaction.inputs.fulfills`
+    - `id` in transaction payload
+    - `id` in `/assets?search={text_search}` (this is also the asset's id!)
+    - `ID` in prose text
+- Asset id:
+    - `asset_id` in query parameters and Event Stream API
+    - `id` in `transaction.asset` payload
+    - `id` in `/assets?search={text_search}` (this is also the transaction's id!)
+- Fulfillment id:
+    - `type_id` in `transaction.outputs.condition.details`
+- Block id:
+    - `block_id` in query parameters and Event Stream API
+    - `id` in the block payload
+
+General inconsistency: `block_id`, `asset_id` and `tx_id` instead of
+`transaction_id`.
+
+
+Even though, this shouldn't be an argument (as we see the HTTP API descoped
+from the underlying implementation):
+
+- [`txid` usage in bdb-server](https://docs.bigchaindb.com/projects/server/en/master/websocket-event-stream-api.html)
+- [`txid` usage in bdb-server](https://docs.bigchaindb.com/projects/server/en/master/websocket-event-stream-api.html)
+
+
+##### Proposal 1
+
+- Call any identifier that points to an object externally is a `object_id`
+    - e.g. `transaction.inputs.fulfills.tx_id` points to an external tx, hence
+      `_`
+- Call any identifier within the data (describing the data) an `id`
+    - `transaction.id`, points to itself, hence no need for redundancy in it's
+      name
+
+Unfortunately, these rules still leave inconsistencies that are hard to
+resolve. Assuming we could agree on `asset_id` and `tx_id` easily, they could
+be a nice way forward, as it would mean that we don't have to change a lot of
+code:
+- `transaction.inputs.fulfills.txid` ==> `transaction.inputs.fulfills.tx_id`
+- `id` in /assets?search={text_search} ==> `tx_id` or `asset_id`, as it's the
```

Comment:
The `id` in the asset is both the asset_id and the transaction_id that create the asset (same problem as in the comment below).

Another thing to consider is how the data is stored in the underlying database. We sometimes let the database itself enforce that a key is unique and rethinkdb only enforces uniqueness of the primary key. That is why we use `id` so much.

This is not necessarily a problem since we can specify the primary for a table in rethinkdb and mongodb allows us to enforce uniqueness in any index. This is something to be aware if we decide to change the name of some keys.